autogluon/scheduler/reporter.py

```diff
@@ -64,6 +64,9 @@ def __call__(self, **kwargs):
             raise AutoGluonEarlyStop('Stopping!')
 
     def fetch(self, block=True):
+        if not self._queue.qsize():
```

Comment:
@zhanghang1989 
It appears this is the key part of the fix, are you able to verify this is safe?

Reply:
Thanks @jwmueller  for reminding. This change is not safe. Basically it breaks all the search algorithms except for random search. Because we need the blocking call of `self._queue.get()` to wait for next trial finishes.

Reply:
@zhanghang1989 thanks for review. 

As far as I understand the logic is the following: for each training task we have a thread which itself opens a process, please correct we if I am wrong. Then we collect data in Queue from child process and pass to main process. So the issue occurs when we try to call `queue.get`, but the child process is already died due to some reason (for example user passed wrong parameters to particular model). At that moment we have a deadlock.  I assume in order to solve it, there will be a need to change some logic in the multiprocessing / multithreading part of the code. With this change I just avoid calling `queue.get` in order to not have the deadlock. And it is not blocking algorithms, since it will try to call `queue.get` until it will get something from `queue` or the job is done.. If you look at changes in searchers, you will see that it is in `while` loop. I checked it for `skopt` and it works.

@Innixma please see the minimal code example, which will cause deadlock, without these changes. 

Tabular Prediction: 

```
import autogluon as ag
from autogluon.task import TabularPrediction as task

import sklearn.datasets
import pandas as pd

cancer_data = sklearn.datasets.load_breast_cancer(return_X_y=False)
data = pd.DataFrame(cancer_data['data'], columns=cancer_data['feature_names'])
data['target'] = cancer_data['target']

hp_tune = True
cat_options = {
    'num_leaves': ag.space.Int(lower=26, upper=66, default=36),
    'depth': ag.space.Int(lower=4, upper=10, default=6),
}
gbm_options = {
    'num_boost_round': 100,
    'num_leaves': ag.space.Int(lower=26, upper=66, default=36),
}

hyperparameters = {
    'CAT': cat_options,
    'GBM': gbm_options,
}

time_limits = 2 * 60
num_trials = 3
output_directory = './agModels--tuningCancerDataset'
data = task.Dataset(data)
label_column = 'target'

# case 1: random search
search_strategy = 'random'
print('\n' * 2)
print('#' * 100)
print(f'Case 1: Running HPO suing search strategy = "{search_strategy}".')
print()
predictor = task.fit(
    train_data=data,
    holdout_frac=0.3,
    label=label_column,
    output_directory=output_directory,
    time_limits=time_limits,
    num_trials=num_trials,
    hyperparameter_tune=hp_tune,
    hyperparameters=hyperparameters,
    search_strategy=search_strategy,
    auto_stack=False,
    stack_ensemble_levels=0
)

print('\n' * 2)
print('#' * 100)
print('HPO results:')
print(predictor.leaderboard(silent=True))

# case 2: skopt search
search_strategy = 'skopt'
print('\n' * 2)
print('#' * 100)
print(f'Case 2: Running HPO suing search strategy = "{search_strategy}".')
print()
predictor = task.fit(
    train_data=data,
    holdout_frac=0.3,
    label=label_column,
    output_directory=output_directory,
    time_limits=time_limits,
    num_trials=num_trials,
    hyperparameter_tune=hp_tune,
    hyperparameters=hyperparameters,
    search_strategy=search_strategy,
    auto_stack=False,
    stack_ensemble_levels=0
)
print('\n' * 2)
print('#' * 100)
print('HPO results:')
print(predictor.leaderboard(silent=True))
```
Here I put wrong parameter `num_leaves` for `catboost`: `catboost` does not support changing it when you run on CPU.

Image Classification: 
```
import autogluon as ag
from autogluon import ImageClassification as task

train_data = task.Dataset(name='fashionmnist')

classifier = task.fit(train_data,
                      num_trials=1,
                      net=None,
                      epochs=5,
                      ngpus_per_trial=1,
                      verbose=False,
                      auto_search=False)
```
Here I put wrong parameter `net=None` .

Reply:
OK, I see now. @zhanghang1989 

I think this IS a problem. With this insertion, a call to fetch() becomes non-blocking, right? Before, it seems that fetch() would not return anything (and wait) until somethere is there to return (please confirm).

But now, it immediately returns None if nothing is there to report, and this is not what we want. This would really break things for the schedulers.

To be clear, there are three different things that could happen when call fetch():
A) Nothing new to be reported
B) Something new to be reported, but not a final metric at the end of training
C) Something new to be reported, namely final metric

Currently, schedulers assume that for A), reporter.fetch() blocks, until B) or C). And scheduler then has to distinguish between them.

Is this blocking the scheduler from doing other things? No, because we have a reporter thread, and only that is blocked (it runs _run_reporter)

Reply:
What if the process is killed, which is supposed to return something in queue ? Then we have deadlock.